runtime/sema/type.go

```diff
@@ -7825,6 +7837,12 @@ func checkSubTypeWithoutEquality(subType Type, superType Type) bool {
 					IsSubsetOf(typedSubType.EffectiveInterfaceConformanceSet())
 			}
 
+			// STRINGERTODO: other options? how to make existing simple types
+			// conform to an intersection type?
+			if typedSuperType.Types[0].Identifier == StringerTypeName {
+				return IsStringerType(subType)
+			}
+
```

Comment:
Not sure if this applies since `StringerType` is an interface and from my understanding it only makes sense to check an interface conformance as part of an intersection type, not directly as a subtype (e.g. `var x: Stringer = 1` is invalid).

Reply:
The `checkSubTypeWithoutEquality` method is a recursive function, so for each type in an interface set `{I1, I2, Stringer}` (or a simple case where `{Stringer}`), it will call the same method recursively, by each interface-type as the super-type. So it makes sense to have `Stringer` type at the top level of the switch-case. Any regular interface types are also at the top level of the switch case, see [sema/type.go#L7927](https://github.com/onflow/cadence/blob/aa514550ad5eca99db4f43bd99f643f4ef79bdad/runtime/sema/type.go#L7927).

> (e.g. var x: Stringer = 1 is invalid).

This is a separate check done at [a different place](https://github.com/onflow/cadence/blob/aa514550ad5eca99db4f43bd99f643f4ef79bdad/runtime/sema/checker.go#L2340). It prevents someone from using an interface type outside of a interface-set/intersection type.

Reply:
Currently, moving the check to the top level of the switch-case does not work, perhaps I'm misunderstanding but it does not seem to call itself recursively once it enters [sema/type.go#L7724](https://github.com/onflow/cadence/blob/aa514550ad5eca99db4f43bd99f643f4ef79bdad/runtime/sema/type.go#L7724).

Reply:
Oh I see the problem.
I think then we can extend the changes you already added/mentioned in https://github.com/onflow/cadence/pull/3594#discussion_r1785154740, to also other non-simple types like `AddressType` and `NumberType`.

i.e: introduce a new Go interface, something like

```golang
type ConformingType interface {
	Type
	EffectiveInterfaceConformanceSet() *InterfaceSet
}
```
And make sure `SimpleType`, `AddressType` and `NumberType` all implement the new interface `ConformingType`.

Then in then [sema/type.go#L7776](https://github.com/onflow/cadence/blob/aa514550ad5eca99db4f43bd99f643f4ef79bdad/runtime/sema/type.go#L7776), add a new `case ConformingType:`, and there, perform the subset check, similar to the composite type (i.e: `typedSuperType.EffectiveIntersectionSet().IsSubsetOf(typedSubType.EffectiveInterfaceConformanceSet())`). Might be able to even combine the two, by making `CompositeType` also implement `ConformingType` at Go level.